### PR TITLE
[issue 75] hadoop-connectors examples

### PR DESCRIPTION
**Pull request by yangb8**
Signed-off-by: Bo Yang <b.yang@emc.com>

This PR adds a number of examples that exercise the Pravega Input Format implementation in the hadoop connectors project.

The files below are copied from apache hadoop and modfied a little bit to support Pravega
hadoop-examples/src/main/java/io/pravega/examples/hadoop/ExampleDriver.java
hadoop-examples/src/main/java/io/pravega/examples/hadoop/RandomTextWriter.java
hadoop-examples/src/main/java/io/pravega/examples/hadoop/RandomWriter.java
hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordCount.java
hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMean.java
hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMedian.java

### CODE REVIEW

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/ExampleDriver.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/**
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/ExampleDriver.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/RandomWriter.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/RandomWriter.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordCount.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordCount.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMean.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMean.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMedian.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordMedian.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/spark/WordCount.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/spark/WordCount.java]: accepted. must be modified by my IDE unintentionally.

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/PravegaOutputFormat.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/PravegaOutputFormat.java]: accepted. make it same as other source files

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/PravegaOutputFormat.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An OutputFormat that can be added as a storage to write events to Pravega.
+ */
+public class PravegaOutputFormat<V> extends OutputFormat<String, V> {
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/PravegaOutputFormat.java]: This PravegaOutputFormat is just developed to write testing data to pravega. it isn't good enough to be published as formal API to be used by end users. I opened a new issue (https://github.com/pravega/hadoop-connectors/issues/28) to implement a generic OutputFormat for hadoop-connectors. Once done, I can replace PravegaOutputFormat in this sample code by the one in hadoop-connectors.

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordCount.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+import org.apache.hadoop.util.GenericOptionsParser;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.util.StringTokenizer;
+
+import com.google.common.base.Charsets;
+
+public class WordCount {
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/WordCount.java]: accepted

File: hadoop-examples/README.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
+# hadoop-connectors examples (Under Development)
```
[the remark this answers is not in the document]
Comment (yangb8) [reply hadoop-examples/README.md]: accepted

**Comment by yangb8** [comment]
this PR is to add examples of hadoop-connectors, so end users can check the source code and run commands listed in README to understand how to use this connector to setup/run MR jobs.

https://github.com/pravega/pravega-samples/issues/75

File: hadoop-examples/src/main/java/io/pravega/examples/hadoop/ExampleDriver.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fpj) [reply hadoop-examples/src/main/java/io/pravega/examples/hadoop/ExampleDriver.java]: I'm sorry, but the issue of copying licensed work is a bit delicate. We need to comply with the following terms of the license:

https://github.com/apache/hadoop/blob/trunk/LICENSE.txt#L90

My underdstanding is that we are allowed to change the license header in the way you did and to add a copyright statement according to the last paragraph of the clause and from paragraph (a) we are providing a copy of the Apache license in our distribution of the work. 

For paragraph (b), we are saying that we have modified it, but I'd make sure to say more explicitly where we are copying from. Put down the whole path and perhaps even the github url.

For paragraph (c), it says that we need to retain any _copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works_. If there is any copyright statement in the original file or in the hadoop/NOTICE.txt file, then we need to add to our own NOTICE file. I scanned the hadoop/NOTICE.txt and nothing stood out for me, so I think we are good.

For paragraph (d), I didn't see anything about the code you are copying, but you might want to have a second look to double check.

**Comment by yangb8** [comment]
@fpj thanks for your response in details.

I totally agree with a) b) d) and I will add full URLs for these copied files, however, I am not quite sure about (c). 
> You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works; 

the header in the original file is like "Licensed to the Apache Software Foundation (ASF) under one * or more contributor license agreements... ...". Shouldn't we keep the original header to comply with (c)?

File: hadoop-examples/README.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+
+# hadoop-connectors examples
+Examples of Hadoop Connectors for Pravega.
+
+Description
+-----------
+These examples give you some basic ideas how to use hadoop-connectors for pravega
```
Comment (RaulGracia) [inline hadoop-examples/README.md]: Please, add a final dot in the sentence.

File: hadoop-examples/README.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+Run Examples on Local Machine
+---
+
+
+Hadoop (verified with Hadoop 2.8.3 on Ubuntu 16.04)
+```
+1. setup and start hdfs
+
+2. set env variables
+   export HDFS=hdfs://192.168.0.188:9000
```
Comment (RaulGracia) [inline hadoop-examples/README.md]: Maybe a `localhost` or `your_ip` would be clearer than an arbitrary IP address.

File: hadoop-examples/README.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+   export HADOOP_EXAMPLES_JAR=build/libs/pravega-hadoop-examples-0.3.0-SNAPSHOT-all.jar
```
Comment (RaulGracia) [inline hadoop-examples/README.md]: I suggest to make this path as absolute, given that following the example flow if the user goes to the another directory (e.g., Spark home) the jar will not be found.

File: hadoop-examples/README.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+Additionally, you can run WordCount program (more will be coming soon) on top of [HiBench](https://github.com/intel-hadoop/HiBench)
+```
+0. set same env variables as previous section, and
+   export HADOOP_HOME=~/services/hadoop-2.8.3
+   export HDFS=hdfs://192.168.0.188:9000
+   export INPUT_HDFS="${HADOOP_EXAMPLES_INPUT_DUMMY} ${PRAVEGA_URI} ${PRAVEGA_SCOPE} ${PRAVEGA_STREAM}"
+
+1. fetch/build/patch HiBench
+   gradle wcHiBench
```
Comment (RaulGracia) [inline hadoop-examples/README.md]: Executing this from scratch, I got HiBench 7.0. With this version, I got several error messages asking for more environment variables apart from these ones (HADOOP_EXECUTABLE, HADOOP_CONF_DIR, WORKLOAD_RESULT_FOLDER, DATASIZE, NUM_MAPS, NUM_REDS, SLAVES). Have you experienced that with HiBench 7.0?

File: hadoop-examples/README.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+2. prepare testing data
+   ./HiBench/bin/workloads/micro/wordcount/prepare/prepare.sh
```
Comment (RaulGracia) [inline hadoop-examples/README.md]: Unfortunately, I could not make it run with HiBench 7.0 at the moment (looks like a configuration problem). Perhaps it is a problem of my environment, but the aspect to consider is that I just followed the instructions (other users might experience similar issues). If I have any advance in this regard, I will provide feedback to you.

File: hadoop-examples/README.md (hunk 11 above)
Comment (yangb8) [reply hadoop-examples/README.md]: accepted.

File: hadoop-examples/README.md (hunk 12 above)
Comment (yangb8) [reply hadoop-examples/README.md]: accepted

File: hadoop-examples/README.md (hunk 13 above)
Comment (yangb8) [reply hadoop-examples/README.md]: accepted

File: hadoop-examples/README.md (hunk 14 above)
Comment (yangb8) [reply hadoop-examples/README.md]: No, I've never met the similar issue.  Do you mind to send the output to me? so I can check if I can reproduce it at local? thanks

**Comment by fpj** [comment]
@yangb8 about your question on c), it says that:

```
          You must retain, in the Source form of any Derivative Works
          that You distribute, all copyright, patent, trademark, and
          attribution notices from the Source form of the Work,
          excluding those notices that do not pertain to any part of
          the Derivative Works
```

It talks about `copyright, patent, trademark, and attribution notices`. The license header is none of that. The license header says, among other things, that the software is licensed to the ASF. Based on the last paragraph of the section:

```
      You may add Your own copyright statement to Your modifications and
      may provide additional or different license terms and conditions
      for use, reproduction, or distribution of Your modifications, or
      for any such Derivative Works as a whole, provided Your use,
      reproduction, and distribution of the Work otherwise complies with
      the conditions stated in this License. 
```

We are allowed to provide our own licensing terms, as long as we don't ignore any of  `copyright, patent, trademark, and attribution notices`, which I don't think there is any.

**Comment by yangb8** [comment]
@fpj thanks for the reply. are there any other comments for this PR?

**Pull request merged**